helios-testing/src/main/java/com/spotify/helios/testing/HeliosSoloDeployment.java

```diff
@@ -122,29 +123,100 @@
+    this.namespace = randomString();
     this.env = containerEnv(builder.env);
     this.binds = containerBinds();
 
     final String heliosPort;
+    final int watchdogPort;
     try {
       final String heliosHost = determineHeliosHost(dockerInfo);
       this.heliosContainerId = deploySolo(heliosHost);
       heliosPort = getHostPort(this.heliosContainerId, HELIOS_MASTER_PORT);
+      watchdogPort = Integer.parseInt(getHostPort(
+          this.heliosContainerId, HELIOS_SOLO_WATCHDOG_PORT));
     } catch (HeliosDeploymentException e) {
       throw new AssertionError("Unable to deploy helios-solo container.", e);
     }
 
-    // Running the String host:port through HostAndPort does some validation for us.
-    this.deploymentAddress = HostAndPort.fromString(dockerHost.address() + ":" + heliosPort);
+    final URI uri = URI.create("http://" + dockerHost.address() + ":" + heliosPort);
+    this.deploymentAddresses = singleton(uri);
     this.heliosClient = Optional.fromNullable(builder.heliosClient).or(
         HeliosClient.newBuilder()
-            .setUser(username)
-            .setEndpoints("http://" + deploymentAddress)
+            .setUser(randomString())
+            .setEndpoints(uri)
             .build());
+
+    watchdogSocket = connectToWatchdogWithTimeout(dockerHost.address(), watchdogPort, 60,
+                                                  TimeUnit.SECONDS, builder.soloWatchdogConnector);
+
+    try {
+      // TODO (dxia) Should we have the constructor start up the container?
+      // Or should do something like extending an abstract class of
+      // com.google.common.util.concurrent.Service?
+      start();
+    } catch (TimeoutException e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  /**
+   * Return a new socket that's connected to the watchdog process of {@link HeliosSoloDeployment}.
+   * @param host The hostname of the Docker host.
+   * @param port The watchdog port to connect to.
+   * @param timeout Timeout
+   * @param timeUnit The {@link TimeUnit}
+   * @return {@link Socket}
+   */
+  @VisibleForTesting
+  static Socket connectToWatchdogWithTimeout(final String host, final int port,
+                                             final long timeout, final TimeUnit timeUnit,
+                                             final SoloWatchdogConnector connector) {
+    final Socket socket;
+
+    try {
+      socket = Polling.awaitUnchecked(timeout, timeUnit, new Callable<Socket>() {
+        @Override
+        public Socket call() throws Exception {
+          final Socket socket = new Socket();
+          try {
+            connector.connect(socket, host, port);
+            return socket;
+          } catch (IOException e) {
+            return null;
+          }
+        }
+      });
+    } catch (TimeoutException e) {
+      throw Throwables.propagate(e);
+    }
+
+    log.info("Connecting to helios-solo watchdog at {}:{}", host, port);
```

Comment:
if the connection is already made at this point, the log line should really be "*Connected* to helios-solo watchdog at ..."